File: typescript/packages/otel/src/span.ts
import type { KnownEvent } from "@threads/core/internal/feed";

// A span while the walk builds it, and the closed span it becomes (spec/otel/README.md).

export type AttrValue = string | number | boolean | readonly string[];
export type Attrs = Record<string, AttrValue | undefined>;

export const INTERNAL = 1;
export const CLIENT = 3;

export type SpanEvent = {
  readonly time: number;
  readonly name: string;
  readonly attributes: Attrs;
};

export type Link = { readonly traceId: string; readonly spanId: string };

/** A closed span: what spans() returns and the encoder writes. Times are ms. */
export type Span = Link & {
  readonly branchId: string;
  readonly closeSeq: number;
  readonly parentSpanId: string | undefined;
  readonly name: string;
  readonly kind: typeof INTERNAL | typeof CLIENT;
  readonly start: number;
  readonly end: number;
  readonly attributes: Attrs;
  readonly events: readonly SpanEvent[];
  readonly links: readonly Link[];
  readonly status: string | undefined;
};

/** Where a span sits: its trace, and the span it is a child of. */
export type Context = {
  readonly traceId: string;
  readonly parentSpanId: string | undefined;
};

/** A span the walk has opened and not yet closed. */
export class OpenSpan {
  readonly events: SpanEvent[] = [];
  readonly attrs: Attrs;

  constructor(
    readonly id: Link,
    readonly parentSpanId: string | undefined,
    readonly name: string,
    readonly kind: typeof INTERNAL | typeof CLIENT,
    readonly opener: KnownEvent,
    attrs: Attrs,
    readonly links: readonly Link[] = [],
  ) {
    this.attrs = { ...attrs };
  }

  /** The span as `close` ends it, for the branch `branchId` exports. */
  close(
    close: KnownEvent,
    branchId: string,
    tenant: string,
    status: string | undefined,
  ): Span {
    const skew = close.time < this.opener.time;
    return {
      ...this.id,
      branchId,
      closeSeq: close.seq,
      parentSpanId: this.parentSpanId,
      name: this.name,
      kind: this.kind,
      start: this.opener.time,
      end: skew ? this.opener.time : close.time,
      attributes: {
        ...this.attrs,
        "threads.tenant": tenant,
        "threads.thread_id": this.opener.thread_id,
        "threads.branch_id": branchId,
        "threads.seq.start": this.opener.seq,
        "threads.seq.end": close.seq,
        "threads.clock_skew": skew || undefined,
      },
      events: this.events,
      links: this.links,
      status,
    };
  }
}

File: python/src/threads/otel/span.py
"""A span while the walk builds it, and the closed span it becomes (spec/otel/README.md)."""

from dataclasses import dataclass, field
from typing import Final

from threads.log import Event

type AttrValue = str | int | bool | tuple[str, ...]
type Attrs = dict[str, AttrValue | None]

INTERNAL: Final = 1
CLIENT: Final = 3


@dataclass(frozen=True, slots=True)
class SpanEvent:
    time: int
    name: str
    attributes: Attrs


@dataclass(frozen=True, slots=True)
class Link:
    trace_id: str
    span_id: str


@dataclass(frozen=True, slots=True)
class Span:
    """A closed span: what `spans()` returns and the encoder writes. Times are ms."""

    trace_id: str
    span_id: str
    branch_id: str
    close_seq: int
    parent_span_id: str | None
    name: str
    kind: int
    start: int
    end: int
    attributes: Attrs
    events: tuple[SpanEvent, ...]
    links: tuple[Link, ...]
    status: str | None


@dataclass(frozen=True, slots=True)
class Context:
    """Where a span sits: its trace, the span it is a child of, and whether its parent thread was
    missing (a child's first turn)."""

    trace_id: str
    parent_span_id: str | None
    parent_missing: bool = False
    run_id: str | None = None
    """The user_input that opened the run, when a user_input opened it."""


@dataclass(slots=True)
class OpenSpan:
    """A span the walk has opened and not yet closed."""

    id: Link
    parent_span_id: str | None
    name: str
    kind: int
    opener: Event
    attrs: Attrs
    links: tuple[Link, ...] = ()
    events: list[SpanEvent] = field(default_factory=list[SpanEvent])

    def close(self, close: Event, branch_id: str, tenant: str, status: str | None) -> Span:
        """The span as `close` ends it, for the branch `branch_id` exports."""
        skew = close.time < self.opener.time
        attributes: Attrs = {
            **self.attrs,
            "threads.tenant": tenant,
            "threads.thread_id": self.opener.thread_id,
            "threads.branch_id": branch_id,
            "threads.seq.start": self.opener.seq,
            "threads.seq.end": close.seq,
            "threads.clock_skew": skew or None,
        }
        return Span(
            self.id.trace_id,
            self.id.span_id,
            branch_id,
            close.seq,
            self.parent_span_id,
            self.name,
            self.kind,
            self.opener.time,
            self.opener.time if skew else close.time,
            attributes,
            tuple(self.events),
            self.links,
            status,
        )
